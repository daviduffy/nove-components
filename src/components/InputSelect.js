// External Dependencies
import React from 'react';
import PropTypes from 'prop-types';

// Internal Dependencies
import { titleize } from '../utils/utils';

const InputSelect = ({
  id,
  label,
  classes,
  error,
  value,
  options,
  onChange,
  style,
  type, // purposely stripped and not used
  required,
  ...rest
}) => {
  const inputClasses = [];
  // TODO: maybe find a better way to do this
  if (value === '') {
    inputClasses.push('F__i--is-unset');
  }
  if (error) {
    inputClasses.push('F__i--err');
  }
  return (
    <div className={`F__g${classes ? ` ${classes}` : ''}`} style={style}>
      {
        label &&
        <label htmlFor={id} className={`F__l${required ? ' F__l--req' : ''}`}>
          <span>{label}</span>
        </label>
      }
      <select
        className={`F__i F__i--s ${inputClasses.join(' ')}`}
        id={id}
        value={value}
        onChange={onChange}
        {...rest}
      >
        {
          options && options.map((option) => {
            let value, text;
            if (typeof (option) === 'string') {
              value = option.toUpperCase();
              text = titleize(option);
            } else {
              [value, text] = option;
            }
            return (
              <option key={value} value={value}>{text}</option>
            );
          })
        }
      </select>
    </div>
  );
};
InputSelect.propTypes = {
  classes: PropTypes.string,
  error: PropTypes.bool,
  id: PropTypes.string,
  label: PropTypes.string,
  options: PropTypes.oneOfType([
    PropTypes.bool,
    PropTypes.array
  ]).isRequired,
  onChange: PropTypes.func.isRequired,
  required: PropTypes.bool,
  style: PropTypes.object,
  value: PropTypes.string
};
InputSelect.defaultProps = {
  classes: '',
  error: false,
  id: '',
  label: '',
  required: false,
  style: {},
  value: ''
};

export default InputSelect;
