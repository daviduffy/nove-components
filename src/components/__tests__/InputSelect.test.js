// External Dependencies
import React from 'react';
import { shallow } from 'enzyme';

// Internal Dependencies
import InputSelect from '../InputSelect';
import { select as selectProps } from '../../fixtures/input';

let wrapper;
beforeEach(() => {
  wrapper = shallow(<InputSelect {...selectProps} />);
});

test('should render InputSelect correctly', () => {
  expect(wrapper).toMatchSnapshot();
});

test('should not render label in InputSelect', () => {
  const noLabelProps = {
    ...selectProps,
    label: ''
  };
  wrapper = shallow(<InputSelect {...noLabelProps} />);
  expect(wrapper.find('label').exists()).toEqual(false);
});

test('should render with error class when error is present', () => {
  expect(wrapper.find('.F__i--err')).toHaveLength(0);
  const errorProps = {
    ...selectProps,
    error: true
  };
  wrapper = shallow(<InputSelect {...errorProps} />);
  expect(wrapper.find('.F__i--err')).toHaveLength(1);
});

test('should render with required class when required', () => {
  expect(wrapper.find('.F__l--req')).toHaveLength(0);
  const requiredProps = {
    ...selectProps,
    required: true
  };
  wrapper = shallow(<InputSelect {...requiredProps} />);
  expect(wrapper.find('.F__l--req')).toHaveLength(1);
});

test('should not allow type attribute even if type supplied', () => {
  expect(wrapper.find('select')).toHaveLength(1);
  expect(wrapper.find('[type="text"]')).toHaveLength(0);
  const numberProps = {
    ...selectProps,
    type: 'text'
  };
  wrapper = shallow(<InputSelect {...numberProps} />);
  expect(wrapper.find('[type="text"]')).toHaveLength(0);
});
