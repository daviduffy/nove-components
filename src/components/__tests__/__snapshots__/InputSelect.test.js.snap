// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render InputSelect correctly 1`] = `
<div
  className="F__g F__g--name"
  style={Object {}}
>
  <label
    className="F__l"
    htmlFor="message"
  >
    <span>
      Input Label
    </span>
  </label>
  <select
    autoFocus={false}
    className="F__i F__i--s F__i--is-unset"
    id="message"
    onChange={[Function]}
    value=""
  >
    <option
      key="AL"
      value="AL"
    >
      Alabama
    </option>
    <option
      key="AK"
      value="AK"
    >
      Alaska
    </option>
    <option
      key="AZ"
      value="AZ"
    >
      Arizona
    </option>
    <option
      key="AR"
      value="AR"
    >
      Arkansas
    </option>
    <option
      key="CA"
      value="CA"
    >
      California
    </option>
    <option
      key="CO"
      value="CO"
    >
      Colorado
    </option>
    <option
      key="CT"
      value="CT"
    >
      Connecticut
    </option>
    <option
      key="DE"
      value="DE"
    >
      Delaware
    </option>
    <option
      key="FL"
      value="FL"
    >
      Florida
    </option>
    <option
      key="GA"
      value="GA"
    >
      Georgia
    </option>
    <option
      key="HI"
      value="HI"
    >
      Hawaii
    </option>
    <option
      key="ID"
      value="ID"
    >
      Idaho
    </option>
    <option
      key="IL"
      value="IL"
    >
      Illinois
    </option>
    <option
      key="IN"
      value="IN"
    >
      Indiana
    </option>
    <option
      key="IA"
      value="IA"
    >
      Iowa
    </option>
    <option
      key="KS"
      value="KS"
    >
      Kansas
    </option>
    <option
      key="KY"
      value="KY"
    >
      Kentucky
    </option>
    <option
      key="LA"
      value="LA"
    >
      Louisiana
    </option>
    <option
      key="ME"
      value="ME"
    >
      Maine
    </option>
    <option
      key="MD"
      value="MD"
    >
      Maryland
    </option>
    <option
      key="MA"
      value="MA"
    >
      Massachusetts
    </option>
    <option
      key="MI"
      value="MI"
    >
      Michigan
    </option>
    <option
      key="MN"
      value="MN"
    >
      Minnesota
    </option>
    <option
      key="MS"
      value="MS"
    >
      Mississippi
    </option>
    <option
      key="MO"
      value="MO"
    >
      Missouri
    </option>
    <option
      key="MT"
      value="MT"
    >
      Montana
    </option>
    <option
      key="NE"
      value="NE"
    >
      Nebraska
    </option>
    <option
      key="NV"
      value="NV"
    >
      Nevada
    </option>
    <option
      key="NH"
      value="NH"
    >
      New Hampshire
    </option>
    <option
      key="NJ"
      value="NJ"
    >
      New Jersey
    </option>
    <option
      key="NM"
      value="NM"
    >
      New Mexico
    </option>
    <option
      key="NY"
      value="NY"
    >
      New York
    </option>
    <option
      key="NC"
      value="NC"
    >
      North Carolina
    </option>
    <option
      key="ND"
      value="ND"
    >
      North Dakota
    </option>
    <option
      key="OH"
      value="OH"
    >
      Ohio
    </option>
    <option
      key="OK"
      value="OK"
    >
      Oklahoma
    </option>
    <option
      key="OR"
      value="OR"
    >
      Oregon
    </option>
    <option
      key="PA"
      value="PA"
    >
      Pennsylvania
    </option>
    <option
      key="RI"
      value="RI"
    >
      Rhode Island
    </option>
    <option
      key="SC"
      value="SC"
    >
      South Carolina
    </option>
    <option
      key="SD"
      value="SD"
    >
      South Dakota
    </option>
    <option
      key="TN"
      value="TN"
    >
      Tennessee
    </option>
    <option
      key="TX"
      value="TX"
    >
      Texas
    </option>
    <option
      key="UT"
      value="UT"
    >
      Utah
    </option>
    <option
      key="VT"
      value="VT"
    >
      Vermont
    </option>
    <option
      key="VA"
      value="VA"
    >
      Virginia
    </option>
    <option
      key="WA"
      value="WA"
    >
      Washington
    </option>
    <option
      key="WV"
      value="WV"
    >
      West Virginia
    </option>
    <option
      key="WI"
      value="WI"
    >
      Wisconsin
    </option>
    <option
      key="WY"
      value="WY"
    >
      Wyoming
    </option>
  </select>
</div>
`;
